### PR TITLE
[CARBONDATA-3705] Support create and load MV for spark datasource table

### PR DESCRIPTION
**Pull request by jackylk**
This PR is based on #3612 

 ### Why is this PR needed?
Materialized View is a feature built on top of Spark, it should support not only carbondata format but also other formats.
 
 ### What changes were proposed in this PR?
1. Added an API in CarbonMetaStore to lookup any relation, not just carbon relation
2. When creating MV, use newly added lookup relation to get all parent table relations. Use CatalogTable instead of CarbonTable whenever possible
3. Skip the segment handling when loading MV
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/317/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2019/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/318/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2020/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/339/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2041/

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala [hunk 1]
```diff
@@ -104,6 +104,58 @@ class MVCreateTestCase extends QueryTest with BeforeAndAfterAll {
     sql(s"""LOAD DATA local inpath '$resourcesPath/data_big.csv' INTO TABLE fact_table6 OPTIONS('DELIMITER'= ',', 'QUOTECHAR'= '"')""")
   }
 
+  test("test create mv on parquet spark table") {
+    sql("drop materialized view if exists mv1")
+    sql("drop table if exists source")
+    sql("create table source using parquet as select * from fact_table1")
+    sql("create materialized view mv1 as select empname, deptname, avg(salary) from source group by empname, deptname")
+    val df = sql("select empname, avg(salary) from source group by empname")
```
Comment (Indhumathi27) [inline datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: Please add a testcase with following steps:
1. create parquet table, create mv, chekc results
2. load to parquet table
3. check results

Comment (jackylk) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: ok, added

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/400/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2101/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/440/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2140/

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 2]
```diff
@@ -568,17 +573,18 @@ object MVHelper {
     generatePartitionerField(allPartitionColumn.toList, Seq.empty)
   }
 
-  private def inheritTablePropertiesFromMainTable(parentTable: CarbonTable,
+  private def inheritTablePropertiesFromMainTable(
```
Comment (akashrn5) [inline mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: this method we can skip for non carbon tables right?

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 3]
```diff
@@ -568,17 +573,18 @@ object MVHelper {
...
+      parentTable: CarbonTable,
       fields: Seq[Field],
       fieldRelationMap: scala.collection.mutable.LinkedHashMap[Field, MVField],
       tableProperties: mutable.Map[String, String]): Unit = {
     var neworder = Seq[String]()
-    val parentOrder = parentTable.getSortColumns.asScala
+    val parentOrder = parentTable.getSortColumns().asScala
```
Comment (akashrn5) [inline mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: please revert the change as old one is correct

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 4]
```diff
@@ -220,55 +209,55 @@ object MVHelper {
...
-      dmProperties("path")
+    val mvTablePath = if (properties.contains("path")) {
+      properties("path")
     } else {
-      CarbonEnv.getTablePath(tableModel.databaseNameOp, tableModel.tableName)(sparkSession)
+      CarbonEnv.getTablePath(mvTableModel.databaseNameOp, mvTableModel.tableName)(sparkSession)
     }
-    CarbonCreateTableCommand(TableNewProcessor(tableModel),
-      tableModel.ifNotExistsSet, Some(tablePath), isVisible = false).run(sparkSession)
+    CarbonCreateTableCommand(TableNewProcessor(mvTableModel),
```
Comment (akashrn5) [inline mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: since we can give sort columns and other things in properties during create of noncarbon table, can we have tests some of them for Materialize views created for non carbon tables?

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 5]
```diff
@@ -568,17 +573,18 @@ object MVHelper {
...
       tableProperties: mutable.Map[String, String]): Unit = {
     var neworder = Seq[String]()
-    val parentOrder = parentTable.getSortColumns.asScala
+    val parentOrder = parentTable.getSortColumns().asScala
     parentOrder.foreach(parentcol =>
       fields.filter(col => fieldRelationMap(col).aggregateFunction.isEmpty &&
                            fieldRelationMap(col).columnTableRelationList.size == 1 &&
                            parentcol.equalsIgnoreCase(
-                             fieldRelationMap(col).columnTableRelationList(0).parentColumnName))
+                             fieldRelationMap(col).columnTableRelationList(0).columnName))
```
[the remark this answers is not in the document]
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: fixed

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 6]
```diff
@@ -610,13 +616,12 @@ object MVHelper {
...
-            relationList.size == 1 &&
-            longStringColumnInParents.contains(relationList.head.parentColumnName)) {
-          varcharDatamapFields += relationList.head.parentColumnName
+        if (aggFunc.isEmpty && relationList.size == 1 && longStringColumnInParents
+          .contains(relationList.head.columnName)) {
+          varcharDatamapFields += relationList.head.columnName
         }
       })
-      if (varcharDatamapFields.nonEmpty) {
+      if (!varcharDatamapFields.isEmpty) {
```
[the remark this answers is not in the document]
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: fixed

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala [hunk 7]
```diff
@@ -72,42 +69,41 @@ object MVHelper {
...
-    val modularPlan = validateMVQuery(sparkSession, logicalPlan)
+    val modularPlan = validateMVQuery(sparkSession, queryPlan)
     val updatedQueryWithDb = modularPlan.asCompactSQL
-    val (timeSeriesColumn, granularity): (String, String) = validateMVTimeSeriesQuery(
-      logicalPlan,
-      dataMapSchema)
-    val fullRebuild = isFullReload(logicalPlan)
+    val (timeSeriesColumn, granularity) = validateMVTimeSeriesQuery(queryPlan, dataMapSchema)
+    val isQueryNeedFullRebuild = checkIsQueryNeedFullRebuild(queryPlan)
+    val isHasNonCarbonProvider = checkIsMainTableHasNonCarbonSource(mainTables)
```
[the remark this answers is not in the document]
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: fixed

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala (hunk 2 above)
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: yes, fixed

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala (hunk 3 above)
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: fixed

File: mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala [hunk 8]
```diff
@@ -104,6 +104,79 @@ class MVCreateTestCase extends QueryTest with BeforeAndAfterAll {
...
+    // load to parquet table and check again
+    sql("insert into source select * from fact_table1")
+    df = sql("select empname, avg(salary) from source group by empname")
+    assert(TestUtil.verifyMVDataMap(df.queryExecution.optimizedPlan, "mv1"))
+    checkAnswer(df, sql("select empname, avg(salary) from fact_table2 group by empname"))
+
+    sql(s"drop materialized view mv1")
+    sql("drop table source")
+  }
+
```
[the remark this answers is not in the document]
Comment (jackylk) [reply mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: added

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala (hunk 4 above)
Comment (jackylk) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: Do you mean providing sort_columns property when creating the MV? If user put sort_columns property in parent table (parquet table),  carbon MV should not inherit it, right?

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/457/

File: mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala (hunk 4 above)
Comment (akashrn5) [reply mv/core/src/main/scala/org/apache/carbondata/mv/extension/MVHelper.scala]: yes

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2158/

**Comment by QiangCai** [comment]
LGTM
